components/backup-stream/src/router.rs

```diff
@@ -2413,4 +2424,77 @@ mod tests {
         let r = cfg_manager.dispatch(changed);
         assert!(r.is_err());
     }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+    async fn test_flush_on_events_race() -> Result<()> {
+        let (tx, _rx) = dummy_scheduler();
+        let tmp = std::env::temp_dir().join(format!("{}", uuid::Uuid::new_v4()));
+        let router = Arc::new(RouterInner::new(
+            tx,
+            Config {
+                prefix: tmp.clone(),
+                // disable auto flush.
+                temp_file_size_limit: 1000,
+                temp_file_memory_quota: 2,
+                max_flush_interval: Duration::from_secs(300),
+            },
+        ));
+
+        let (task, _path) = task("race".to_owned()).await?;
+        must_register_table(router.as_ref(), task, 1).await;
+        router
+            .must_mut_task_info("race", |i| {
+                i.storage = Arc::new(NoopStorage::default());
+            })
+            .await;
+        let mut b = KvEventsBuilder::new(42, 0);
+        b.put_table(CF_DEFAULT, 1, "k1".as_bytes(), "v1".as_bytes());
+        let events_before_flush = b.finish();
+
+        b.put_table(CF_DEFAULT, 1, "k1".as_bytes(), "v1".as_bytes());
+        let events_after_flush = b.finish();
+
+        // make timestamp precision to 1 seconds.
+        fail::cfg("temp_file_name_timestamp", "return(1000)").unwrap();
+
+        let (trigger_tx, trigger_rx) = std::sync::mpsc::sync_channel(0);
+        let trigger_rx = std::sync::Mutex::new(trigger_rx);
+
+        let t = router.get_task_info("race").await.unwrap();
+        let _ = router.on_events(events_before_flush).await;
+
+        // make generate temp files ***happen after*** moving files to flushing_files in
+        // do_flush
+        fail::cfg_callback("before_generate_temp_file", move || {
+            trigger_rx.lock().unwrap().recv().unwrap();
+        })
+        .unwrap();
+
+        fail::cfg_callback("after_moving_to_flushing_files", move || {
```

Comment:
Just a note here: I think it will be good to add a happens-before framework like rocksdb did, the interface might be:
```rust
fail_utils::sequentially("failpoint1", "falipoint2", ctx);
```
The context type may looks like (May be wrapped for safely running concurrently.):
```rust
struct SeqCtx {
  // failpoint name -> how many its predecessor called or events awaiting for its predecessor happen.
  semaphore: Map<String, Semaphore>,
  // the casual order.
  happens_before: Map<String, String>
}
```
When calling `fail_utils::sequentially(F1, F2, ctx)`, it:
1. Set semaphore[F2] to a zero-permit semaphore.

2. Add the following callback to F1:
    1. query the context to check whether there are any events happens AFTER it, namely F2. 
    2. get the semaphore of them, add a permit to it.

3. Then, the following callback will be added to F2:
    1. try to acquire a semaphore of F2.
    2. forget (but not return) the semaphore permit.

Reply:
cool

Reply:
Then I'd like to use sync_channel to implement it until `fail-rs` support this. don't need to reinvent the wheel here.